azurelinuxagent/common/utils/processutil.py

```diff
@@ -0,0 +1,226 @@
+
+
+def capture_from_process_pre_33(process, cmd, timeout):
+    """
+    Can't use process.communicate(timeout=), so do it the hard way.
+    """
+    def kill_on_timeout(pid, watcher_timeout):
+        """
+        Check for the continued existence of pid once per second. If pid no longer exists, exit with code 0.
+        If timeout (in seconds) elapses, kill pid and exit with code 1.
+        """
+        for iteration in range(watcher_timeout):
+            time.sleep(1)
+            try:
+                os.kill(pid, 0)
+            except OSError as ex:
+                if 3 == ex.errno:   # Process no longer exists
+                    exit(0)
+        os.killpg(os.getpgid(pid), signal.SIGKILL)
+        exit(1)
+
+    watcher = Process(target=kill_on_timeout, args=(process.pid, timeout))
+    watcher.start()
+
+    try:
+        # Now, block "forever" waiting on process. If the timeout-limited Event wait in the watcher pops,
+        # it will kill the process and Popen.communicate() will return
+        stdout, stderr = process.communicate()
+    except OSError as e:
+        _destroy_process(process, signal.SIGKILL)
+        raise ExtensionError("Error while running '{0}': {1}".format(cmd, e.strerror))
+    except RuntimeError as e:
+        _destroy_process(process, signal.SIGKILL)
+        raise ExtensionError("RuntimeError while running '{0}': {1}".format(cmd, e))
+    except Exception as e:
+        _destroy_process(process, signal.SIGKILL)
+        raise ExtensionError("Exception while running '{0}': {1}".format(cmd, e))
+
+    timeout_happened = False
+    watcher.join(1)
+    if watcher.is_alive():
+        watcher.terminate()
+    else:
+        timeout_happened = (watcher.exitcode == 1)
+
+    if timeout_happened:
+        msg = format_stdout_stderr(sanitize(stdout), sanitize(stderr))
+        raise ExtensionError("Timeout({0}): {1}\n{2}".format(timeout, cmd, msg))
+
+    return stdout, stderr
+
+
+def capture_from_process_no_timeout(process, cmd):
+    try:
+        stdout, stderr = process.communicate()
+    except OSError as e:
+        _destroy_process(process, signal.SIGKILL)
+        raise ExtensionError("Error while running '{0}': {1}".format(cmd, e.strerror))
+    except Exception as e:
+        _destroy_process(process, signal.SIGKILL)
+        raise ExtensionError("Exception while running '{0}': {1}".format(cmd, e))
+
+    return stdout, stderr
+
+
+def capture_from_process_raw(process, cmd, timeout):
+    """
+    Captures stdout and stderr from an already-created process.
+
+    :param subprocess.Popen process: Created by subprocess.Popen()
+    :param str cmd: The command string to be included in any exceptions
+    :param int timeout: Number of seconds the process is permitted to run
+    :return: The stdout and stderr captured from the process
+    :rtype: (str, str)
+    :raises ExtensionError: if a timeout occurred or if anything was raised by Popen.communicate()
+    """
+    if not timeout:
```

Comment:
not timeout seems weird.  Does this mean timeout == 0 or timeout <= 0 or timeout < 0 or something else?

Reply:
I think explicit is better than conversion.  YMMV.